### ISSUE TITLE
Confusion around `spin add` / Have a directory for every spin component

### ISSUE THREAD

**Issue by calebschoepp** [fermyon/spin#1075]
**TL;DR** `spin new` should nest its component in a directory to make `spin add` less confusing.

When I was trying the `spin add` command for the first time it did not work like I expected it to. I was adding an `http-rust` component (called "test") to an existing `rust-http` application. My expectation was that some new code would be added to my existing `src/` directory. Instead a new `test/` directory was generated with it's own enclosed `test/src/`. I know understand that this was because we need separate web assembly components and thus separate cargo projects, but it violated my expectations.

After discussing this with @michelleN I've concluded that the root issue here is that we don't have a directory for every spin component. Put another way, we don't nest the initial component created by `spin new` in a directory of its own, we only do this for components made by `spin add`.

I think that this creates a poor user experience when you consecutively run `spin new` and `spin add`. You'll end up with a file hierarchy something like this:

```
my_project/
    src/
    Cargo.toml
    spin.toml
    component_i_added/
        src/
        Cargo.toml
```

I don't think any user actually wants there project organized that way. Instead I think that they would prefer to have it organized as each component living in a directory. Rather than forcing the user to understand how to reorganize the above we could just put the first component from `spin new` in its own directory giving us the below after running `spin new` and `spin add`:

```
my_project/
    component_from_new/
        src/
        Cargo.toml
    component_from_add/
        src/
        Cargo.toml
    spin.toml    
```

This is a pretty big change so my goal with this issue is just to spur some discussion.

**Comment by itowlson** [comment]
The original thinking (okay, maybe "thinking" is too strong a word) was that for a single-component app it's a little weird for it to have the additional level of directory structure - imagine the examples each having their actual code tucked away in a subdirectory.  It is inconsistent with the `spin add` (multi-component) behaviour, but arguably does provide for a less jarring experience in the simple case.  But yes, it does make for a poor structure if the user creates a non-empty app and then adds to it.  (If they plan ahead, they can use `http-empty` and add to that, but many apps grow more organically than that!)

I agree it would be "nicer" in some sense to add code to the existing codebase and have builds that produce the different components - this would make it much easier to share code.  But it also means that without savvy build tricks, which would be highly language dependent, you'd likely get everything built in everywhere.  Keeping the components totally separate at least has the benefit of being simple and obvious.

**Comment by calebschoepp** [comment]
> was that for a single-component app it's a little weird for it to have the additional level of directory structure - imagine the examples each having their actual code tucked away in a subdirectory.

I totally agree that this is a problem with the sub-directory for every component from the start approach. A bit of a tradeoff. I suppose if users typically only use one component than it is not worth it. But, maybe if typical user applications have lots of components then maybe it is.

> I agree it would be "nicer" in some sense to add code to the existing codebase and have builds that produce the different components - this would make it much easier to share code. 

Agreed it would be nice, but to be clear I'm not advocating in this issue that we should try to do these savy build tricks. I think a sub-directory for every component is adequate.
